library/sha3.c

```diff
@@ -0,0 +1,390 @@
+        s[22] ^= (~lane[3]) & lane[4];
+        s[23] ^= (~lane[4]) & lane[0];
+        s[24] ^= (~lane[0]) & lane[1];
+
+        /* Iota */
+        s[0] ^= rc[round];
+    }
+}
+
+static uint8_t left_encode( uint8_t *encbuf, size_t value )
+{
+    uint8_t n = 0;
+
+    /* Compute length of value in bytes */
+    for ( size_t v = value; v > 0 && n < sizeof( size_t ); n++, v >>= 8 );
+
+    if ( n == 0 )
+        n = 1;
+
+    encbuf[0] = ( uint8_t )n;
+    for ( int i = 1; i <= n; i++ )
+        encbuf[i] = ( uint8_t )( value >> ( 8 * ( n - i ) ) );
+
+    return( n + 1 );
+}
+
+static void keccak_pad( mbedtls_sha3_context *ctx )
+{
+    if( ( ctx->index % ctx->max_block_size ) != 0 )
+    {
+        uint8_t b = 0;
+        ctx->index = ctx->r / 8 - 1;
+        mbedtls_sha3_update( ctx, &b, 1 );
+    }
+}
+
+void mbedtls_sha3_init( mbedtls_sha3_context *ctx )
+{
+    if( ctx == NULL )
+        return;
+
+    memset( ctx, 0, sizeof( mbedtls_sha3_context ) );
+}
+
+void mbedtls_sha3_free( mbedtls_sha3_context *ctx )
+{
+    if( ctx == NULL )
+        return;
+
+    mbedtls_platform_zeroize( ctx, sizeof( mbedtls_sha3_context ) );
+}
+
+void mbedtls_sha3_clone( mbedtls_sha3_context *dst,
+                                    const mbedtls_sha3_context *src )
+{
+    if ( dst == NULL || src == NULL )
+        return;
+
+    *dst = *src;
+}
+
+/*
+ * SHA-3 context setup
+ */
+int mbedtls_sha3_starts( mbedtls_sha3_context *ctx, mbedtls_sha3_id id )
+{
+    mbedtls_sha3_family_functions *p = NULL;
+    if( ctx == NULL )
+        return( MBEDTLS_ERR_SHA3_BAD_INPUT_DATA );
+
+    for( p = sha3_families; p->id != MBEDTLS_SHA3_NONE; p++ )
+    {
+        if( p->id == id )
+            break;
+    }
+
+    if( p == NULL )
+        return( MBEDTLS_ERR_SHA3_BAD_INPUT_DATA );
+
+    ctx->id = id;
+    ctx->r = p->r;
+    ctx->olen = p->olen / 8;
+    ctx->xor_byte = p->xor_byte;
+    ctx->max_block_size = ctx->r / 8;
+
+    return( 0 );
+}
+
+/*
+ * SHA-3 starts with name and customization (for CSHAKE)
+ */
+/* If this function receives an id != CSHAKE, it fallsback to mbedtls_sha3_starts() */
+int mbedtls_sha3_starts_cshake( mbedtls_sha3_context *ctx, mbedtls_sha3_id id,
+                                const uint8_t *name, size_t name_len,
+                                const uint8_t *custom, size_t custom_len )
+{
+    int ret = 0;
+    size_t encbuf_len = 0;
+    uint8_t encbuf[sizeof( size_t ) + 1];
+
+    if( ( ( name == NULL || name_len == 0 ) &&
+        ( custom == NULL || custom_len == 0 ) ) ||
+        ( id != MBEDTLS_SHA3_CSHAKE128 && id != MBEDTLS_SHA3_CSHAKE256 ) )
+    {
+        if( id == MBEDTLS_SHA3_CSHAKE128 )
+            return( mbedtls_sha3_starts( ctx, MBEDTLS_SHA3_SHAKE128 ) );
+        else if( id == MBEDTLS_SHA3_CSHAKE256 )
+            return( mbedtls_sha3_starts( ctx, MBEDTLS_SHA3_SHAKE256 ) );
+        /* If other id is provided, silently start the context */
+        return( mbedtls_sha3_starts( ctx, id ) );
+    }
+
```

Comment:
If the `id` is one of the CSHAKE variants, and one of the names is null but the other is not, this if won't be triggered, we could pass the null or zero length on. The update function does handle either of these cases, so no invalid accesses will happen, but passing a null pointer, with a non-zero length would result in nonsensical hash output.

Reply:
Fixed. If a parameter is NULL, its len is set to 0 and the parameter is not sha3_updated (only the len).